src/test/java/eu/europeana/metis/sandbox/service/workflow/HarvestServiceImplTest.java

```diff
@@ -1,239 +1,336 @@
+        assertEquals(Country.NETHERLANDS, captorRecordInfo.getValue().getRecord().getCountry());
+        assertEquals(Language.NL, captorRecordInfo.getValue().getRecord().getLanguage());
+        assertEquals("datasetName", captorRecordInfo.getValue().getRecord().getDatasetName());
     }
 
     @Test
-    void harvestOaiRecordHeader_ExpectSuccess() throws HarvesterException {
+    void harvestOaiPmh_exceedingLimitWithXslt_expectSuccess() throws HarvesterException {
+        harvestService = new HarvestServiceImpl(httpHarvester, oaiHarvester, asyncRecordPublishService, datasetService, 1, recordRepository);
+        OaiHarvestData oaiHarvestData = new OaiHarvestData("url", "setspec", "metadaformat", "oaiIdentifier");
 
-        harvestService = new HarvestServiceImpl(httpHarvester, oaiHarvester, 1000, recordRepository);
-        RecordEntity recordEntity = new RecordEntity(null, null, "1");
-        OaiHarvestData oaiHarvestData = new OaiHarvestData("someEndpointURL", "someSetSpec", "somePrefix", "someOaiIdentifier");
-        Record.RecordBuilder record = Record.builder()
-                .datasetId("1")
-                .country(Country.NETHERLANDS)
-                .language(Language.NL);
+        OaiRecordHeaderIterator oaiRecordHeaderIterator = new TestUtils.TestHeaderIterator(prepareListForOaiRecordIterator());
 
-        OaiRecordHeader recordHeader = new OaiRecordHeader("someId", false, Instant.now());
-        OaiRecord oaiRecord = new OaiRecord(recordHeader,
-                () -> "record".getBytes(StandardCharsets.UTF_8));
+        OaiRecord mockOaiRecord = mock(OaiRecord.class);
+        RecordEntity recordEntity = new RecordEntity("", "","datasetId");
+        recordEntity.setId(1L);
 
-        when(oaiHarvester.harvestRecord(any(OaiRepository.class), anyString())).thenReturn(oaiRecord);
+        when(oaiHarvester.harvestRecordHeaders(any(OaiHarvest.class))).thenReturn(oaiRecordHeaderIterator);
+        when(oaiHarvester.harvestRecord(any(OaiRepository.class), anyString())).thenReturn(mockOaiRecord);
+        when(mockOaiRecord.getRecord()).thenReturn(new ByteArrayInputStream("record".getBytes(StandardCharsets.UTF_8)));
+        when(datasetService.isXsltPresent(anyString())).thenReturn(true);
         when(recordRepository.save(any(RecordEntity.class))).thenReturn(recordEntity);
 
-        var harvestContent = harvestService.harvestOaiRecordHeader("1", oaiHarvestData, record);
-
-        assertEquals("1", harvestContent.getRecord().getDatasetId());
-        assertEquals(Country.NETHERLANDS, harvestContent.getRecord().getCountry());
-        assertEquals(Language.NL, harvestContent.getRecord().getLanguage());
-        assertTrue(CollectionUtils.isEmpty(harvestContent.getErrors()));
-        assertEquals("record", new String(harvestContent.getRecord().getContent(), StandardCharsets.UTF_8));
+        harvestService.harvestOaiPmh("datasetId", createMockEncapsulatedRecord(), oaiHarvestData);
 
```

Comment:
remove duplicate code assertions can be extracted to a method. ```assertHarvestProcess```

```
        verify(asyncRecordPublishService, times(XXX)).publishWithoutXslt(captorRecordInfo.capture(), eq(Step.XXX));
        verify(datasetService).updateNumberOfTotalRecord(eq("datasetId"), eq(XXX));
        assertTrue(captorRecordInfo.getAllValues().stream().allMatch(x -> x.getRecord().getDatasetId().equals("datasetId")));
        assertTrue(captorRecordInfo.getAllValues().stream().allMatch(x -> x.getRecord().getContent() != null));
        assertEquals(Country.NETHERLANDS, captorRecordInfo.getValue().getRecord().getCountry());
        assertEquals(Language.NL, captorRecordInfo.getValue().getRecord().getLanguage());
        assertEquals("datasetName", captorRecordInfo.getValue().getRecord().getDatasetName());
```

Reply:
done